pyuvdata/uvcal/uvcal.py

```diff
@@ -3351,6 +3355,446 @@ def _convert_to_filetype(self, filetype):
+            True or `cal_type="delay"` and if freq_range is not None and has multiple
+            spectral windows, ignored otherwise. Defaulted to uvdata.spw_array if
+            either `wide_band` is True or `cal_type="delay"` and if freq_range is None.
+        jones : array_like of int, optional
+            Calibration Jones elements. If None, defaults to [-5, -6] (jxx, jyy) if
+            uvdata is in linear pol. [-1, -2] (jrr, jll) if uvdata is in circular pol.
+            A ValueError is raised if jones_array is None and uvdata is in
+            psuedo-stokes.
+        ref_antenna_name : str, optional
+            Phase reference antenna, required if cal_style = "sky".
+        sky_catalog : str, optional
+            Name of calibration catalog, required if cal_sky = "sky".
+        sky_field : str, optional
+            Short string describing field center or dominant source, required if
+            cal_sky = "sky".
+        diffuse_model : str, optional
+            Name of diffuse model.
+        baseline_range : array_like of float, optional
+            Range of baselines used for calibration.
+        Nsources : int, optional
+            Number of sources used.
+        observer : str, optional
+            Name of observer who calculated calibration solutions.
+        gain_scale : str, optional
+            The gain scale of the calibration, which indicates the units of the
+            calibrated visibilities. For example, Jy or K str.
+        git_hash_cal : str, optional
+            Commit hash of calibration software (from git_origin_cal) used to generate
+            solutions.
+        git_origin_cal : str, optional
+            Origin (on github for e.g) of calibration software. Url and branch.
+        extra_keywords : dict, optional
+            Any user supplied extra keywords, type=dict.
+
+        Raises
+        ------
+        ValueError
+            If cal_style is 'sky' and ref_antenna_name, sky_catalog or sky_field are not
+            provided;
+            if freq_array is not None, flex_spw is True and flex_spw_id_array is None;
+            if freq_array and channel_width are None and the uvdata object does not use
+            flexible spectral windows and the uvdata channel width varies;
+            if time_array and integration_time are None and the uvdata integration
+            time varies;
+            if time_array is not None and integration_time is not specified or is the
+            wrong type;
+            if jones_array is None and uvdata is in psuedo-stokes.
+
+        """
+        if not issubclass(type(uvdata), UVData):
+            raise ValueError("uvdata must be a UVData (or subclassed) object.")
+
+        # re-initialize to make sure we have an empty object
+        uvc = cls()
+
+        if cal_type == "gain":
+            uvc._set_gain()
+        elif cal_type == "delay":
+            uvc._set_delay()
```

Comment:
Should we be raising an error if `cal_type` doesn't belong to one of these? (I forget if we support cal objects that are essentially "unknown"...)

Reply:
yes. We're planning to remove the "unknown" one, so I'm going to have it error if it's not "gain" or "delay"